### PR TITLE
add autocomplete support for zsh and fish shells

### PR DESCRIPTION
**Pull request by mjurbanski-reef**

### CODE REVIEW

File: b2/_internal/_cli/autocomplete_install.py [hunk 1]
```diff
@@ -84,43 +101,196 @@ def program_in_path(self) -> bool:
...
+        child.expect(pexpect.EOF)
+    except pexpect.TIMEOUT:
+        logger.warning("Command %r timed out, output: %r", cmd, output.getvalue())
+        child.kill(signal.SIGKILL)
+        return False
+    finally:
+        child.close()
+
+    (logger.debug if child.exitstatus == 0 else logger.warning
+    )("Command %r exited with code %r, output: %r", cmd, child.exitstatus, output.getvalue())
```
Comment (emnoor-reef) [inline b2/_internal/_cli/autocomplete_install.py:292]:
```suggestion
    logger.log(
        logging.DEBUG if child.exitstatus == 0 else logging.WARNING,
        "Command %r exited with code %r, output: %r", cmd, child.exitstatus, output.getvalue()
    )
```

File: b2/_internal/_cli/autocomplete_install.py [hunk 2]
```diff
@@ -84,43 +101,196 @@ def program_in_path(self) -> bool:
...
+        stdout, stderr = p.communicate(timeout=timeout)
+    except subprocess.TimeoutExpired:
+        p.kill()
+        stdout, stderr = p.communicate(timeout=1)
+        logger.warning("Command %r timed out, stdout: %r, stderr: %r", cmd, stdout, stderr)
+    else:
+        (logger.debug if p.returncode == 0 else logger.warning)(
+            "Command %r exited with code %r, stdout: %r, stderr: %r", cmd, p.returncode, stdout,
+            stderr
+        )
```
Comment (emnoor-reef) [inline b2/_internal/_cli/autocomplete_install.py:265]:
```suggestion
        logger.log(
            logging.DEBUG if p.returncode == 0 else logging.WARNING,
            "Command %r exited with code %r, stdout: %r, stderr: %r", cmd, p.returncode, stdout,
            stderr
        )
```

**Review by emnoor-reef — APPROVED** [review]

**Comment by emnoor-reef** [comment]
The first time triggering the autocomplete just prints with completion options. On the second time the auto completion can be selected.
IDK, if it is a argcomplete bug or not.

https://github.com/reef-technologies/B2_Command_Line_Tool/assets/137923473/32fc38ab-9fe1-4de1-a019-3b69e287bf50

**Comment by emnoor-reef** [comment]
Something similar happens on fish. The autocomplete selection actually starts on send Tab. But the options are not left in scrollback buffer like zsh. Not a fish user, so IDK if this behaviour is expected.

https://github.com/reef-technologies/B2_Command_Line_Tool/assets/137923473/0d9c6a48-a388-4267-a62b-67b1bc3921b4

**Comment by mjurbanski-reef** [comment]
> The first time triggering the autocomplete just prints with completion options. On the second time the auto completion can be selected. The printed options from the first time stays in the shell scrollback buffer, which doesn't usually happen on zsh. IDK, if it is a argcomplete bug or not.
> 
>  Screen.Recording.2024-03-28.at.9.31.46.PM.mov

this behavior is due to the sheer number of completions suggested as far as I can tell - I can reproduce it using by trying to complete `git -`

**Comment by emnoor-reef** [comment]
If you have too many suggestions, zsh prompts you if you want to see them or not:

> zsh: do you wish to see all 155 possibilities (155 lines)?

For b2 it doesn't happen.

**Comment by mjurbanski-reef** [comment]
thats weird, it does prompt me
image

**Pull request closed without merging**